rpc_sidecar/src/node_client.rs

```diff
@@ -305,151 +296,127 @@ impl Error {
+    ) -> Result<Framed<TcpStream, BinaryMessageCodec>, AnyhowError> {
+        let disconnected_start = Instant::now();
+        inc_disconnect();
+        error!("node connection closed, will attempt to reconnect");
+        let stream = Self::connect_with_retries(config).await?;
+        info!("connection with the node has been re-established");
+        observe_reconnect_time(disconnected_start.elapsed());
+        Ok(stream)
+    }
 }
 
 #[async_trait]
-impl NodeClient for JulietNodeClient {
+impl NodeClient for FramedNodeClient {
     async fn send_request(&self, req: BinaryRequest) -> Result<BinaryResponseAndRequest, Error> {
-        let payload = encode_request(&req).expect("should always serialize a request");
-        let request_guard = self
-            .client
-            .read()
-            .await
-            .create_request(ChannelId::new(0))
-            .with_payload(payload.into())
-            .queue_for_sending()
-            .await;
-        let response = request_guard
-            .wait_for_response()
+        let _permit = self
+            .request_limit
+            .acquire()
             .await
-            .map_err(|err| Error::RequestFailed(err.to_string()))?
-            .ok_or(Error::NoResponseBody)?;
-        let resp = bytesrepr::deserialize_from_slice(&response)
+            .map_err(|err| Error::RequestFailed(err.to_string()))?;
+
+        let payload =
+            BinaryMessage::new(encode_request(&req).expect("should always serialize a request"));
+        let mut client = self.client.write().await;
+
+        if let Err(err) = tokio::time::timeout(
+            Duration::from_secs(self.config.message_timeout_secs),
+            client.send(payload),
+        )
+        .await
+        .map_err(|_| Error::RequestFailed("timeout".to_owned()))?
+        {
+            warn!(
+                addr = %self.config.address,
+                err = display_error(&err),
+                "binary port client handler error"
+            );
+            return Err(Error::RequestFailed(err.to_string()));
+        };
+
+        let Ok(maybe_response) = tokio::time::timeout(
+            Duration::from_secs(self.config.message_timeout_secs),
+            client.next(),
+        )
+        .await
+        else {
+            return Err(Error::RequestFailed("timeout".to_owned()));
+        };
+
+        if let Some(response) = maybe_response {
+            let resp = bytesrepr::deserialize_from_slice(
+                response
+                    .map_err(|err| Error::RequestFailed(err.to_string()))?
+                    .payload(),
+            )
             .map_err(|err| Error::EnvelopeDeserialization(err.to_string()))?;
-        handle_response(resp, &self.shutdown)
+            handle_response(resp, &self.shutdown)
+        } else {
+            let stream = Self::reconnect(&self.config)
+                .await
+                .map_err(|_| Error::RequestFailed("reconnection failed".to_string()))?;
```

Comment:
this will make the caller wait for a re-connect, which IMO is unnecessary, I think we should maybe just close the connection here so that the re-connect loop picks it up? or notify the re-connect loop somehow

Reply:
it might be worth it to introduce a tiny bit of complexity to signal errors, so that we don't have to re-connect every time something goes wrong

Reply:
Reconnection logic has been rewritten in 96e2a3ec5c696d2c98a49e01886b10668b888f5d